compiler/runtime/memory.c

```diff
@@ -28,24 +34,40 @@ size_t sk_pow2_size(size_t size) {
   return (1 << sk_bit_size(size));
 }
 
-void sk_add_ftable(void* ptr, size_t size) {
-  size_t slot = sk_bit_size(size);
-  *(void**)ptr = sk_ftable[slot];
-  sk_ftable[slot] = ptr;
+size_t sk_obj_size(size_t size, sk_size_info_t* size_info) {
+  size_t pow2_size = sk_pow2_size(size);
+  size_t bit_size = sk_bit_size(size);
+  if (pow2_size > size && bit_size > 8) {
+    size_t main_size = 1 << (bit_size - 1);
+    size_t reminder_size = sk_pow2_size(size - main_size);
+    if (main_size != reminder_size) {
+      size_info->main = bit_size - 1;
+      size_info->remainder = sk_bit_size(reminder_size);
+      return main_size + reminder_size;
+    }
+  }
+  size_info->main = bit_size;
+  size_info->remainder = 0;
+  return  pow2_size;
 }
 
-void* sk_get_ftable(size_t size) {
-  size_t slot = sk_bit_size(size);
-  void** ptr = sk_ftable[slot];
+void sk_add_ftable(void* ptr, sk_size_info_t* si) {
```

Comment:
This is a bit of a micro optimisation, but I think you're better off passing si by value here (and in other places such as sk_get_ftable where it's only read). The compiler will just utilise registers for the two size_t values, which si will be loaded in to anyway, and save a potential tlb or l1 miss. But I think it opens up inlining too.

Reply:
>  But I think it opens up inlining too.

This was based on past experience, not anything the compiler must take care about, such as aliasing. I just checked out of curiosity. It will help gdb to inline. Clang 15 figures it out regardless.